limit the containers that can be executed

logan-wrld opened the pull request:
### Review Type Requested (choose one):
- [ ] **Logic** - thorough check (from everybody doing review)

### Summary
Limit the containers that can be executed on the jobcreator. Updates consist of a module allowlist fetch and allowlist check before deal agreement. 

### Task/Issue reference
Closes: #63

### Details 

Implemented into the jobcreator/controller.go due to controller managing job offers and doing the interaction with the solver on deal agreements. This includes the allowlist fetch using a simple http fetch to the lilypad-tech Github as well as the allowlist check function before a deal is agreed on. 

### How to test this code? (optional)
I am still having issues with my lilypad on-chain jobs in my local dev environment- please test as well!

### Anything else? (optional)
I did a alias to the http package as to not step on the "github.com/lilypad-tech/lilypad/pkg/http" package

logan-wrld committed:
fetch allowlist and allowlist check before deal agreement

logan-wrld committed:
fixed one http request issue, now fix the git request

logan-wrld committed:
added lilypad/pkg/http back

logan-wrld committed:
alias to http package for git request

logan-wrld committed:
allowlist dealstate callout

bgins commented on pkg/jobcreator/controller.go:
```diff
@@ -190,6 +220,56 @@ func (controller *JobCreatorController) Start(ctx context.Context, cm *system.Cl
+	allowlistUrl := "https://raw.githubusercontent.com/lilypad-tech/module-allowlist/master/allowlist.txt"
+
+	// Create a new HTTP request with a context
+	req, err := httpAlias.NewRequest("GET", allowlistUrl, nil)
```
We might want to add a new helper function to the `"github.com/lilypad-tech/lilypad/pkg/http"` package to handle this case and others that need a timeout. We could create a `GetRequestWithTimeout` that takes the necessary request parameters and a timeout argument.

The idea is to keep all of our HTTP communication going through one package. And it's pretty likely we'll need something like this again in the future!

logan-wrld replied:
I added a GetRequestWithTimeout helper function within the pkg/http/utils.go - HTTP comms are going through one package!

bgins replied:
Awesome! Thanks ❤️

bgins commented on pkg/jobcreator/controller.go:
```diff
@@ -190,6 +220,56 @@ func (controller *JobCreatorController) Start(ctx context.Context, cm *system.Cl
+	controller.web3Events.Storage.SubscribeDealStateChange(func(ev storage.StorageDealStateChange) {
+		deal, err := controller.solverClient.GetDeal(ev.DealId)
+		if err != nil {
+			controller.log.Error("module allolist error", err)
```
```suggestion
			controller.log.Error("module allowlist error", err)
```

bgins commented on pkg/jobcreator/controller.go:
```diff
@@ -190,6 +220,56 @@ func (controller *JobCreatorController) Start(ctx context.Context, cm *system.Cl
 }
+func (controller *JobCreatorController) UpdateModuleAllowlist() error {
+	// URL of the GitHub raw content of the allowlist
+	allowlistUrl := "https://raw.githubusercontent.com/lilypad-tech/module-allowlist/master/allowlist.txt"
```
Could we make this a package level const? If we won't change it at runtime, we could have it as a read-only value.

Also, is this URL correct? Currently reports a 404.

logan-wrld replied:
Makes way more sense than a hardcoded string! The allowlistURL is now sourced from a package-level constant instead of being hard-coded within the function. The module-allowlist is now configured on Github so you should not be seeing a 404 error any longer.

bgins commented on pkg/jobcreator/controller.go:
```diff
@@ -145,8 +149,29 @@ func (controller *JobCreatorController) subscribeToWeb3() error {
+	// Initial fetch of the module allowlist
+	err := controller.UpdateModuleAllowlist()
+	if err != nil {
+		// handle error
```
How do we plan to handle this error?

logan-wrld replied:
I fixed this by passing the error to the controller with contoller.log.Error... Does this work? Especially since this error is mainly for testing, passing it to web3 metrics is overkill?

bgins replied:
Yeah, agreed passing it to metrics probably isn't needed. We'd probably get a bunch of reports whenever someone typos cowsay. 😆

Let's chat about the `controller.log.Error` approach. It's probably right, but would love to chat about it because I am still getting familiar with the code base.

logan-wrld committed:
Update pkg/jobcreator/controller.go

Co-authored-by: Brian Ginsburg <7957636+bgins@users.noreply.github.com>

logan-wrld committed:
Update controller.go

AquiGorka committed:
feat: add geolocation data to websocket events (#147)

* feat: add geolocation data to websocket events

* feat: add deals tracking (#151)

* feat: add function to track deals

* feat: add tracking deals

* chore: fix typo

arsen3d committed:
feat: Configuration update for Arbitrum Local l2 (#142)

* Configuration update for Arbitrum

* chore: remove file not part of PR

* chore: remove settings that are not needed and set global defaults and consts

* chore: refactor stack to simplify local chain clean, run and boot

---------

Co-authored-by: Gorka <gorka@aquigorka.com>

AquiGorka committed:
build: update embeds (#156)

logan-wrld committed:
UpdateModuleAllowlist set as constant

logan-wrld committed:
GetRequestWithTimeout helper function

hunjixin committed:
Feat/pow support (#143)

* feat: power implementation

* fix: blocked error channel

---------

Co-authored-by: root <root@ip-172-31-44-171.ap-southeast-1.compute.internal>

arsen3d committed:
Configuration update for Arbitrum (#154)

AquiGorka committed:
Revert "Feat/pow support (#143)" (#158)

This reverts commit fc3602c35726b5df40507ece6d0b65507da14ee8.

bgins committed:
build: Update configuration and workflows for Devnet (#152)

* build: Rename solver, job creator, and contracts secrets

* chore: Remove unused Cloudflare token and Doppler config

* chore: Rename local_l2 to dev

Matches other places where we refer to local development as dev.

* fix: update network on run cmd

* chore: Update print-env network, RPC URL, and chain ID

* build: Remove Devnet contracts deploy job

We have decided to rebuild the chain from scratch on Devnet.

---------

Co-authored-by: Gorka <gorka@aquigorka.com>

bgins commented on pkg/jobcreator/controller.go:
```diff
@@ -228,6 +290,7 @@ func (controller *JobCreatorController) solve() error {
  *
 */
 
+// list the deals we have been assigned to that we have not yet posted and agree tx to the contract for
```
We can remove this one. It's a duplicate of the line below.

hunjixin committed:
Feat/pow support (#161)

* feat: power implementation

* feat: simplly code

* chore: add a explain comment for target difficulty

* fix typo

* fix: blocked error channel

* chore: update dev toml sc address

* feat: change wrong log and add toml tag

* fix: upgrade cowsay to v0.0.4

* fix: update devent controller address

* fix: Add POW config processor

* feat: add pow address for testnet

* chore: update to latest controller address

* chore: update contract addresses

* build: add hardhat testnet config

---------

Co-authored-by: root <root@ip-172-31-44-171.ap-southeast-1.compute.internal>
Co-authored-by: hunjixin <16705420332lee@gmai.com>
Co-authored-by: Brian Ginsburg <gins@brianginsburg.com>
Co-authored-by: Gorka <gorka@aquigorka.com>
Co-authored-by: Arsen Yeremin <arsen3d@gmail.com>

hunjixin committed:
fix: pow  deploy typo (#162)

Co-authored-by: hunjixin <16705420332lee@gmai.com>

AquiGorka committed:
docs: update local development instructions (#164)

logan-wrld committed:
Added original comments back

logan-wrld committed:
chore: fix formatting
